osgp/platform/osgp-adapter-ws-smartmetering/src/main/java/org/opensmartgridplatform/adapter/ws/smartmetering/endpoints/SmartMeteringConfigurationEndpoint.java

```diff
@@ -1586,4 +1592,72 @@ public SetRandomisationSettingsResponse getSetRandomisationSettingsResponse(
     }
     return response;
   }
+
+  @PayloadRoot(localPart = "GetKeysRequest", namespace = SMARTMETER_CONFIGURATION_NAMESPACE)
+  @ResponsePayload
+  public GetKeysAsyncResponse getKeys(
+      @OrganisationIdentification final String organisationIdentification,
+      @RequestPayload final GetKeysRequest request,
+      @MessagePriority final String messagePriority,
+      @ScheduleTime final String scheduleTime,
+      @ResponseUrl final String responseUrl,
+      @BypassRetry final String bypassRetry)
+      throws OsgpException {
+
+    final GetKeysRequestData dataRequest =
+        this.configurationMapper.map(request.getGetKeysData(), GetKeysRequestData.class);
+
+    final RequestMessageMetadata requestMessageMetadata =
+        RequestMessageMetadata.newBuilder()
+            .withOrganisationIdentification(organisationIdentification)
+            .withDeviceIdentification(request.getDeviceIdentification())
+            .withDeviceFunction(DeviceFunction.GET_KEYS)
+            .withMessageType(MessageType.GET_KEYS)
+            .withMessagePriority(messagePriority)
+            .withScheduleTime(scheduleTime)
+            .withBypassRetry(bypassRetry)
+            .build();
+
+    final AsyncResponse asyncResponse =
+        this.requestService.enqueueAndSendRequest(requestMessageMetadata, dataRequest);
+
+    this.saveResponseUrlIfNeeded(asyncResponse.getCorrelationUid(), responseUrl);
+
+    return this.configurationMapper.map(asyncResponse, GetKeysAsyncResponse.class);
+  }
+
+  @PayloadRoot(localPart = "GetKeysAsyncRequest", namespace = SMARTMETER_CONFIGURATION_NAMESPACE)
+  @ResponsePayload
+  public GetKeysResponse getGetKeysResponse(@RequestPayload final GetKeysAsyncRequest request)
+      throws OsgpException {
+
+    GetKeysResponse response = null;
+    try {
+      response = new GetKeysResponse();
+
+      final ResponseData responseData =
+          this.responseDataService.dequeue(
+              request.getCorrelationUid(), ComponentType.WS_SMART_METERING);
+
+      this.throwExceptionIfResultNotOk(responseData, "Get keys");
+
+      response.setResult(OsgpResultType.fromValue(responseData.getResultType().getValue()));
+
+      if (responseData.getMessageData() != null) {
+        final List<GetKeysResponseData> target = response.getGetKeysResponseData();
+        final org.opensmartgridplatform.domain.core.valueobjects.smartmetering.GetKeysResponse
+            getKeysResponse =
+                (org.opensmartgridplatform.domain.core.valueobjects.smartmetering.GetKeysResponse)
+                    responseData.getMessageData();
+        target.addAll(
+            this.configurationMapper.mapAsList(
+                getKeysResponse.getKeys(), GetKeysResponseData.class));
+      } else {
+        log.info("Get keys response is null");
+      }
+    } catch (final Exception e) {
+      this.handleException(e);
+    }
+    return response;
```

Comment:
`handleException` always throws an exception, so the return statement is never reached. However, the compiler complains about a missing return statement if it is removed. Since the other commands in this class use the same pattern, I keep it like this.